src/porepy/numerics/ad/functions.py

```diff
@@ -141,6 +142,35 @@ def l2_norm(dim: int, var: pp.ad.AdArray) -> pp.ad.AdArray:
     return pp.ad.AdArray(vals, jac)
 
 
+def safe_power(power: float, zero_val: float, var: AdArray) -> AdArray:
+    """Safe (negative) power of an AdArray.
+
+    The power is performed only for nonzeros in the variable, whereas zeros
+    are left as zeros. This is useful for avoiding division by zero in the
+    Jacobian.
+
+    Parameters:
+        numerator: AdArray representing the numerator.
+        denominator: AdArray representing the denominator.
+
+    Returns:
+        AdArray representing the safe division.
+
+    """
+    if isinstance(var, np.ndarray):
```

Comment:
The typing indicates this is not possible. There are of course plenty of cases where our implementation is more general than the typing.

Reply:
This goes for a lot of this module. I think this is necessary because of some fundamentals of the AD functions, but I'm not sure. Let's discuss if you don't recall this.

Reply:
I think you are right. This may be fit for a maintenance project, but I'm also fine with you just resolving the comment.

Reply:
What would you do in that project?

Reply:
That was a revealing question. Touche. I'll resolve.